src/shogun/lib/DynamicArray.h

```diff
@@ -13,30 +13,875 @@
+		 *
+		 * @return array element at last index
+		 */
+		inline bool get_last_element() const
+		{
+			return m_array[num_elements - 1];
+		}
+
+		/** get array element at index
+		 *
+		 * (does bounds checking)
+		 *
+		 * @param index index
+		 * @return array element at index
+		 */
+		inline bool get_element_safe(int32_t index) const
+		{
+			if (index >= get_num_elements())
+			{
+				SG_SERROR(
+				    "array index out of bounds (%d >= %d)\n", index,
+				    get_num_elements());
+			}
+			return m_array[index];
+		}
+
+		/** set array element at index
+		 *
+		 * @param e element to set
+		 * @param idx1 index 1
+		 * @param idx2 index 2
+		 * @param idx3 index 2
+		 * @return if setting was successful
+		 */
+		inline bool set_element(
+		    bool e, int32_t idx1, int32_t idx2 = 0, int32_t idx3 = 0)
+		{
+			int32_t index = idx1 + dim1_size * (idx2 + dim2_size * idx3);
+			if (index < 0)
+			{
+				return false;
+			}
+			else if (index <= num_elements - 1)
+			{
+				m_array[index] = e;
+				return true;
+			}
+			else if (index < array_size)
+			{
+				m_array[index] = e;
+				num_elements = index + 1;
+				return true;
+			}
+			else
+			{
+				if (free_array && resize_array(index))
+					return set_element(e, index);
+				else
+					return false;
+			}
+		}
+
+		/** insert array element at index
+		 *
+		 * @param e element to insert
+		 * @param index index
+		 * @return if setting was successful
+		 */
+		inline bool insert_element(bool element, int32_t index)
+		{
+			if (append_element(get_element(num_elements - 1)))
+			{
+				for (int32_t i = array_size - 2; i > index; i--)
+				{
+					m_array[i] = m_array[i - 1];
+				}
+				m_array[index] = element;
+
+				return true;
+			}
+
+			return false;
+		}
+
+		/** append array element to the end of array
+		 *
+		 * @param e element to append
+		 * @return if setting was successful
+		 */
+		inline bool append_element(bool e)
+		{
+			return set_element(e, array_size);
+		}
+
+		/** STD VECTOR compatible. Append array element to the end
+		 *  of array.
+		 *
+		 * @param e element to append
+		 */
+		inline void push_back(bool e)
+		{
+			if (get_num_elements() < 0)
+				set_element(e, 0);
+			else
+				set_element(e, get_num_elements());
+		}
+
+		/** STD VECTOR compatible. Delete array element at the end
+		 *  of array.
+		 */
+		inline void pop_back()
+		{
+			if (get_num_elements() <= 0)
+				return;
+
+			delete_element(get_num_elements() - 1);
+		}
+
+		/** STD  VECTOR compatible. Return array element at the end
+		 *  of array.
+		 *
+		 * @return element at the end of array
+		 */
+		inline bool back()
+		{
+			if (get_num_elements() <= 0)
+				return get_element(0);
+
+			return get_element(get_num_elements() - 1);
+		}
+
+		/** find first occurence of array element and return its index
+		 * or -1 if not available
+		 *
+		 * @param e element to search for
+		 * @return index of element or -1
+		 */
+		inline int32_t find_element(bool element)
```

Comment:
same as in case of `insert_element`:
```
../src/shogun/lib/DynamicArray.h: In member function ‘int32_t shogun::CDynamicArray<bool>::find_element(bool)’:
../src/shogun/lib/DynamicArray.h:466:3: warning: declaration of ‘element’ shadows a member of 'this' [-Wshadow]
```
use just `e` as function arg var.